pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,302 @@
+	dnsMasq.args = []string{"-k", "-d", "--strict-order", "--bind-dynamic", "--no-resolv", "--dhcp-no-override", "--dhcp-authoritative",
+		"--except-interface=lo", "--dhcp-hostsfile=" + DHCPHostsFile, "--dhcp-leasefile=" + DHCPLeaseFile,
+		"--pid-file=" + DHCPPidFile}
+	OS = &OSHandler{}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+type OSOps interface {
+	IsFileExist(path string) bool
+	getPidFromFile(file string) (int, error)
+	checkProcessExist(pid int) (bool, *os.Process)
+	killProcessIfExist(pid int) error
+	RecreateHostsFile() (*os.File, error)
+	executeCommand(string, []string) (*exec.Cmd, error)
+	readFromFile(file string) ([]byte, error)
+}
+
+type OSHandler struct{}
+
+var OS OSOps
+
+func (o *OSHandler) IsFileExist(path string) bool {
+	_, err := os.Stat(DHCPPidFile)
+	return !os.IsNotExist(err)
+}
+
+func (o *OSHandler) readFromFile(file string) ([]byte, error) {
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to open file: %s", file)
+		return nil, err
+	}
+	return content, nil
+}
+
+func (o *OSHandler) getPidFromFile(file string) (int, error) {
+	content, err := OS.readFromFile(file)
+	if err != nil {
+		return -1, err
+	}
+	lines := strings.Split(string(content), "\n")
+	pid, _ := strconv.Atoi(lines[0])
+	return pid, nil
+}
+
+func (o *OSHandler) killProcessIfExist(pid int) error {
+	procExist, proc := OS.checkProcessExist(pid)
+	if procExist {
+		killErr := proc.Kill()
+		if killErr != nil {
+			log.Log.Warningf("failed to kill process %d", proc.Pid)
+		}
+	}
+	return nil
+}
+
+func (o *OSHandler) checkProcessExist(pid int) (bool, *os.Process) {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		log.Log.Warningf("failed to find process %d", pid)
+		return false, nil
+	}
+	procErr := proc.Signal(syscall.Signal(0))
+	return procErr == nil, proc
+}
+
+func (o *OSHandler) RecreateHostsFile() (*os.File, error) {
+	// Remove the config file
+	ferr := os.RemoveAll(DHCPHostsFile)
+	if ferr != nil {
+		log.Log.Reason(ferr).Warningf("failed to remove dhcp hosts file: %s", DHCPHostsFile)
+	}
+
+	f, err := os.Create(DHCPHostsFile)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to create dnsmasq hosts file in %s", DHCPHostsFile)
+		return nil, err
+	}
+	return f, nil
+}
+
+func (o *OSHandler) executeCommand(path string, startArgs []string) (*exec.Cmd, error) {
+	cmd := exec.Command(path, startArgs...)
+
+	err := cmd.Start()
+	if err != nil {
+		log.Log.Reason(err).Error("failed to start dnsmasq")
+		return nil, err
+	}
+	go func(cmd *exec.Cmd) {
+		err := cmd.Wait()
```

Comment:
if dnsmasq crashes unexpectedly, somehow virt-dhcp needs to know how to recover it.